core/sr-api-macros/tests/runtime_calls.rs

```diff
@@ -114,3 +121,72 @@ fn use_trie_function() {
 	let block_id = BlockId::Number(client.info().unwrap().chain.best_number);
 	assert_eq!(runtime_api.use_trie(&block_id).unwrap(), 2);
 }
+
+#[test]
+fn initialize_block_works() {
+	let client = test_client::new_with_execution_strategy(ExecutionStrategy::Both);
+	let runtime_api = client.runtime_api();
+	let block_id = BlockId::Number(client.info().unwrap().chain.best_number);
+	assert_eq!(runtime_api.get_block_number(&block_id).unwrap(), 1);
+}
+
+#[test]
+fn initialize_block_is_called_only_once() {
+	let client = test_client::new_with_execution_strategy(ExecutionStrategy::Both);
+	let runtime_api = client.runtime_api();
+	let block_id = BlockId::Number(client.info().unwrap().chain.best_number);
+	assert_eq!(runtime_api.take_block_number(&block_id).unwrap(), Some(1));
+	assert_eq!(runtime_api.take_block_number(&block_id).unwrap(), None);
+}
+
+#[test]
+fn initialize_block_is_skipped() {
+	let client = test_client::new_with_execution_strategy(ExecutionStrategy::Both);
+	let runtime_api = client.runtime_api();
+	let block_id = BlockId::Number(client.info().unwrap().chain.best_number);
+	assert!(runtime_api.without_initialize_block(&block_id).unwrap());
+}
+
+#[test]
+fn record_proof_works() {
+	let client = test_client::new_with_execution_strategy(ExecutionStrategy::Both);
+	let mut runtime_api = client.runtime_api();
+
+	let transaction = Transfer {
+		amount: 500,
+		nonce: 0,
+		from: AccountKeyring::Alice.into(),
+		to: Default::default(),
+	}.into_signed_tx();
+
+	// Build the block
+	let mut builder = client.new_block().unwrap();
+	builder.push(transaction.clone()).unwrap();
+	let block = builder.bake().unwrap();
+
+	// Extract some data for further usage
+	let storage_root = block.header().state_root().clone();
+	let block_id = BlockId::hash(block.header().hash());
+
+	// Import the generated block
+	client.import(BlockOrigin::Own, block).unwrap();
+
+	// Generate some proof
+	runtime_api.record_proof();
+
+	assert_eq!(
+		runtime_api.balance_of(
+			&block_id,
+			AccountKeyring::Alice.into()
+		).unwrap(),
+		500,
+	);
+
+	let proof = runtime_api.extract_proof().expect("Should have collected some proof.");
+	assert!(!proof.is_empty());
+
+	create_proof_check_backend::<<<Header as HeaderT>::Hashing as HashT>::Hasher>(
+		storage_root,
+		proof,
+	).expect("Proof contains the storage root of the build block.");
```

Comment:
how about a test re-executing the block with the proof? does the tools for that exist?

Reply:
(that is cumulus :sweat_smile:)

Reply:
I could create a backend out of the proof and call `execute_block` with this backend.

Reply:
> (that is cumulus 😅)

> I could create a backend out of the proof and call `execute_block` with this backend.

Yeah, it would be cool to catch an issue in substrate before it breaks cumulus.